qute.jdt/com.redhat.qute.jdt/src/main/java/com/redhat/qute/jdt/internal/template/datamodel/TemplateGlobalAnnotationSupport.java

```diff
@@ -0,0 +1,206 @@
+		try {
+			IField[] fields = type.getFields();
+			for (IField field : fields) {
+				if (AnnotationUtils.getAnnotation(field, TEMPLATE_GLOBAL_ANNOTATION) == null
+					&& isTemplateGlobalMember(field)) {
+					collectResolversForTemplateGlobal(field, templateGlobal, resolvers, typeResolver, monitor);
+				}
+			}
+			IMethod[] methods = type.getMethods();
+			for (IMethod method : methods) {
+				if (AnnotationUtils.getAnnotation(method, TEMPLATE_GLOBAL_ANNOTATION) == null
+					&& isTemplateGlobalMember(method)) {
+					collectResolversForTemplateGlobal(method, templateGlobal, resolvers, typeResolver, monitor);
+				}
+			}
+		} catch (JavaModelException e) {
+			LOGGER.log(Level.SEVERE, "Error while getting methods of '" + type.getElementName() + "'.", e);
+		}
+	}
+
+	private void collectResolversForTemplateGlobal(IMember member, IAnnotation templateGlobal,
+		List<ValueResolverInfo> resolvers, ITypeResolver typeResolver, IProgressMonitor monitor) {
+		if (isTemplateGlobalMember(member)) {
+			String sourceType = member.getDeclaringType().getFullyQualifiedName();
+			ValueResolverInfo resolver = new ValueResolverInfo();
+			resolver.setSourceType(sourceType);
+			resolver.setSignature(getSignature(member, typeResolver));
+			// Constant value for {@link #name()} indicating that the field/method name
+			// should be used
+			try {
+				resolver.setNamed(
+					AnnotationUtils.getAnnotationMemberValue(templateGlobal, TEMPLATE_GLOBAL_ANNOTATION_NAME));
+			} catch (JavaModelException e) {
+				LOGGER.log(Level.SEVERE, "Error while getting annotation member value of 'named'.", e);
+			}
+			resolver.setGlobalVariable(true);
+			if (!resolvers.contains(resolver)) {
+				resolvers.add(resolver);
+			}
+		}
+	}
+
+	/**
+	 * Returns true if the given member is supported by @TemplateGlobal and false
+	 * otherwise.
+	 *
+	 * A global variable method:
+	 *
+	 * <ul>
+	 * <li>must not be private</li>
+	 * <li>must be static,</li>
+	 * <li>must not accept any parameter,</li>
+	 * <li>must not return {@code void},</li>
+	 * </ul>
+	 *
+	 * A global variable field:
+	 *
+	 * <ul>
+	 * <li>must not be private</li>
+	 * <li>must be static,</li>
+	 * </ul>
+	 *
+	 * @param member the member to check.
+	 * @return true if the given member <code>member</code> is a template global
+	 *         member and false otherwise.
+	 */
+	private static boolean isTemplateGlobalMember(IMember member) {
+		try {
+			// every non-void non-private static method that declares no parameters and
+			// every non-private static field is considered a global variable
+			if (!Modifier.isPrivate(member.getFlags()) && Modifier.isStatic(member.getFlags())) {
+				if (member.getElementType() == IJavaElement.FIELD) {
+					return true;
+				} else if (member.getElementType() == IJavaElement.METHOD) {
+					IMethod method = (IMethod) member;
+					return method.getParameters().length == 0 && !"void".equals(method.getReturnType());

```

Comment:
I would use `method.getNumberOfParameters()`

Also, try defining a void method like :

`public static void doesNothing () {}`

It shows up in the completion when the class is annotated with `@TemplateGlobal` but it shouldn't.

In fact, the same problem exists with `@TemplateExtension`. It doesn't exclude void values.

Reply:
There seems to be an issue with `org.eclipse.jdt.core.Signature`, where it receives a `char[]` of the typeName:

but the resulting string is the first character:

There might be an alternative to `IMethid.getReturnType()` instead?